### PR TITLE
Gen bit perm for Semi honest and malicious setting

### PR DESCRIPTION
**Pull request by richajaindce**
This is stacked on #236

### CODE REVIEW

File: src/protocol/sort/bit_permutation.rs [hunk 1]
```diff
@@ -3,83 +3,93 @@ use std::iter::{repeat, zip};
...
+/// 5. Compute the final output, a vector of length `n`. Each element `i` in this output vector is the sum of
+/// the elements at index `i` and `i+n` from the vector computed in step 4.
+///
+/// ![Bit Permutation steps][bit_permutation]
+/// ## Panics
+/// In case the function is unable to get double size of output from multiplication step, the code will panic
+///
+/// ## Errors
+/// It will propagate errors from multiplication protocol.
+pub async fn bit_permutation<'a, F: Field, S: SecretSharing<F> + Copy>(
```
Comment (richajaindce) [inline src/protocol/sort/bit_permutation.rs:58]: Will try to get rid of this Copy once Martin's changed are landed from #235

File: src/protocol/sort/bit_permutation.rs [hunk 2]
```diff
@@ -3,83 +3,93 @@ use std::iter::{repeat, zip};
...
+    fn share_of_one(&self) -> Self::Share {
+        Replicated::one(self.role())
+    }
+}
+
+impl<F: Field> ShareOfOne<F> for ProtocolContext<'_, MaliciousReplicated<F>, F> {
+    type Share = MaliciousReplicated<F>;
+
+    fn share_of_one(&self) -> Self::Share {
+        MaliciousReplicated::one(self.role(), self.prss().generate_replicated(RECORD_0))
```
Comment (richajaindce) [inline src/protocol/sort/bit_permutation.rs:31]: @benjaminsavage @akoshelev is this the correct way to generate r_share? Context does not hold r_share and thats why I did it like this

File: src/protocol/sort/bit_permutation.rs [hunk 3]
```diff
@@ -3,83 +3,88 @@ use std::iter::{repeat, zip};
...
+pub async fn bit_permutation<'a, F: Field, S: SecretSharing<F> + Clone>(
+    ctx: ProtocolContext<'a, S, F>,
+    input: &[S],
+) -> Result<Vec<S>, BoxError> 
+where ProtocolContext<'a, S, F>: SecureMul<F, Share = S> + ShareOfOne<F, Share = S>{
+    let share_of_one = ctx.share_of_one();
+
+    let mult_input = input
+        .iter()
+        .map(|x: &S| &share_of_one - x)
```
Comment (richajaindce) [inline src/protocol/sort/bit_permutation.rs:65]: cannot subtract `&S` from `&S`. consider extending the `where` clause, but there might be an alternative better way to express this requirement: `, &S: Sub<&S>`

Comment (richajaindce) [reply src/protocol/sort/bit_permutation.rs]: For now, we are cloning share_of_one to take care of this.

File: src/protocol/sort/bit_permutation.rs (hunk 2 above)
Comment (benjaminsavage) [reply src/protocol/sort/bit_permutation.rs:31]: No, this is not correct.

`SecurityValidator` is the thing that knows what the `r_share` is. It has an accessor: `validator.r_share()`.

When the `SecurityValidator` is created via a call to `SecurityValidator::new()` that's when the global value of `r_share` is initialized. 

If you look at the tests in `malicious.rs` like `simplest_circuit` you'll see that right now nobody holds a reference to this thing... that's a problem.

Right now, I just get the `SecurityValidatorAccumulator` from it via a call to `v.accumulator()`, and this object is passed to the context via a call to `upgrade_to_malicious(acc.clone())`

Then, nobody has a reference to the actual `SecurityValidator` ever again... eventually someone needs it in order to call `validate()`

Here is what I'd suggest:

Update the `upgrade_to_malicious()` function to *also* take a clone of `r_share`. Once it has this, you can expose a new `get_share_of_one()` method on the context. Then, the context (who knows both the `role` and `r_share`) will be capable of generating a share of one in the malicious context.

**Review by benjaminsavage — CHANGES_REQUESTED** [review]

File: src/protocol/sort/bit_permutation.rs [hunk 4]
```diff
@@ -3,83 +3,90 @@ use std::iter::{repeat, zip};
...
-impl<'a, F: Field> BitPermutation<'a, F> {
-    pub fn new(input: &'a [Replicated<F>]) -> BitPermutation<'a, F> {
-        Self { input }
+impl<F: Field> ShareOfOne<F> for ProtocolContext<'_, Replicated<F>, F> {
+    type Share = Replicated<F>;
+
+    fn share_of_one(&self) -> Self::Share {
+        Replicated::one(self.role())
     }
+}
```
Comment (benjaminsavage) [inline src/protocol/sort/bit_permutation.rs:25]: Stuff like this should not be defined within this file `bit_permutation.rs` - this is a particular protocol, but many protocols will require a share of one.

File: src/protocol/sort/bit_permutation.rs [hunk 5]
```diff
@@ -3,83 +3,90 @@ use std::iter::{repeat, zip};
...
+/// In case the function is unable to get double size of output from multiplication step, the code will panic
+///
+/// ## Errors
+/// It will propagate errors from multiplication protocol.
+pub async fn bit_permutation<'a, F: Field, S: SecretSharing<F>>(
+    ctx: ProtocolContext<'a, S, F>,
+    input: &[S],
+) -> Result<Vec<S>, BoxError>
+where
+    ProtocolContext<'a, S, F>: SecureMul<F, Share = S> + ShareOfOne<F, Share = S>,
```
Comment (benjaminsavage) [inline src/protocol/sort/bit_permutation.rs:62]: What is this? Do we need to add these types of trait bounds every time we upgrade protocols to use SecretSharing traits?

File: src/protocol/sort/bit_permutation.rs (hunk 4 above)
Comment (richajaindce) [reply src/protocol/sort/bit_permutation.rs:25]: Created a new file called context_trail.rs and moving this code there. Other context related trait definitions can also be added here.

File: src/protocol/sort/bit_permutation.rs (hunk 5 above)
Comment (richajaindce) [reply src/protocol/sort/bit_permutation.rs:62]: @akoshelev is there any better way to get rid of this? This is required by the compiler to know protocol does implement these traits

File: src/protocol/sort/bit_permutation.rs (hunk 2 above)
Comment (richajaindce) [reply src/protocol/sort/bit_permutation.rs:31]: Makes complete sense! Making the change

File: src/protocol/context_traits.rs [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+    fn share_of_one(&self) -> Self::Share {
+        Replicated::one(self.role())
+    }
+}
+
+impl<F: Field> ShareOfOne<F> for ProtocolContext<'_, MaliciousReplicated<F>, F> {
+    type Share = MaliciousReplicated<F>;
+
+    fn share_of_one(&self) -> Self::Share {
+        MaliciousReplicated::one(self.role(), self.r_share())
```
Comment (benjaminsavage) [inline src/protocol/context_traits.rs:25]: this is perfect

File: src/protocol/reveal/mod.rs [hunk 7]
```diff
@@ -177,9 +177,9 @@ mod tests {
 
             let record_id = RecordId::from(i);
             let results = try_join_all(vec![
-                ctx[0].clone().reveal(record_id, &share[0]),
-                ctx[1].clone().reveal(record_id, &share[1]),
-                ctx[2].clone().reveal(record_id, &share[2]),
+                ctx[0].ctx.clone().reveal(record_id, &share[0]),
+                ctx[1].ctx.clone().reveal(record_id, &share[1]),
+                ctx[2].ctx.clone().reveal(record_id, &share[2]),
```
Comment (benjaminsavage) [inline src/protocol/reveal/mod.rs:182]: What is going on here? Why did you need to add a `.ctx`? And what does that even do?

File: src/protocol/sort/bit_permutation.rs [hunk 8]
```diff
@@ -53,13 +55,13 @@ pub async fn bit_permutation<F: Field>(
...
 
-    assert_eq!(mult_output.len(), input.len() * 2);
+    debug_assert!(mult_output.len() == input.len() * 2);
     // Generate permutation location
     let len = mult_output.len() / 2;
     for i in 0..len {
         // we are subtracting "1" from the result since this protocol returns 1-index permutation whereas all other
         // protocols expect 0-indexed permutation
-        let less_one = &mult_output[i + len] - &share_of_one;
+        let less_one = mult_output[i + len].clone() - &share_of_one;
```
Comment (benjaminsavage) [inline src/protocol/sort/bit_permutation.rs:64]: Why did you change this to clone()?

File: src/protocol/sort/bit_permutation.rs [hunk 9]
```diff
@@ -112,4 +117,40 @@ mod tests {
...
+        let [ctx0, ctx1, ctx2] = make_malicious_contexts::<Fp31>(&world);
+        let mut rand = StepRng::new(100, 1);
+
+        let mut shares = [
+            Vec::with_capacity(INPUT.len()),
+            Vec::with_capacity(INPUT.len()),
+            Vec::with_capacity(INPUT.len()),
+        ];
+        for i in INPUT {
+            let share = share_malicious(Fp31::from(*i), &mut rand);
```
Comment (benjaminsavage) [inline src/protocol/sort/bit_permutation.rs:138]: How can you share malicious without knowing the value of `r`?

File: src/protocol/sort/bit_permutation.rs [hunk 10]
```diff
@@ -112,4 +117,40 @@ mod tests {
...
+        let h1_future = bit_permutation(ctx1.ctx, shares[1].as_slice());
+        let h2_future = bit_permutation(ctx2.ctx, shares[2].as_slice());
+
+        let result: [_; 3] = try_join_all([h0_future, h1_future, h2_future])
+            .await
+            .unwrap()
+            .try_into()
+            .unwrap();
+
+        validate_list_of_shares_malicious(EXPECTED, &result);
```
Comment (benjaminsavage) [inline src/protocol/sort/bit_permutation.rs:154]: We should call `validate()` on the SecurityValidator after running the protocol and ensure it is indeed valid.

File: src/protocol/sort/generate_sort_permutation.rs [hunk 11]
```diff
@@ -18,79 +18,58 @@ use embed_doc_image::embed_doc_image;
...
     num_bits: u8,
-}
-
-impl<'a> GenerateSortPermutation<'a> {
-    #[must_use]
-    pub fn new(input: &'a [(u64, u64)], num_bits: u8) -> GenerateSortPermutation {
-        Self { input, num_bits }
-    }
+) -> Result<Vec<Replicated<F>>, BoxError> {
+    let ctx_0 = ctx.narrow(&Sort(0));
```
Comment (benjaminsavage) [inline src/protocol/sort/generate_sort_permutation.rs:41]: Why did you need to add this? Can we remove this?

File: src/protocol/sort/bit_permutation.rs (hunk 9 above)
Comment (benjaminsavage) [reply src/protocol/sort/bit_permutation.rs:138]: Ugh.. I missed this when `share_malicious()` was added to the codebase... I'll fix it in another diff.

File: src/protocol/sort/bit_permutation.rs (hunk 10 above)
Comment (benjaminsavage) [reply src/protocol/sort/bit_permutation.rs:154]: I'll add this in a diff to come.

**Review by benjaminsavage — APPROVED** [review]
This is good work @richajaindce! Thanks for adding this `share_of_one()` functionality!

File: src/protocol/reveal/mod.rs (hunk 7 above)
Comment (richajaindce) [reply src/protocol/reveal/mod.rs:182]: make malicious context returns both context and validator (so that it is not dropped). Thats why we need to explicitly call ctx here

File: src/protocol/sort/bit_permutation.rs (hunk 8 above)
Comment (richajaindce) [reply src/protocol/sort/bit_permutation.rs:64]: We are now performing S + &S vs &S + &S since addition of two &S isnt supported as yet.
